### ISSUE TITLE
TAGS Question / Feature Request

### ISSUE THREAD

**Issue by Andre188** [balabit/syslog-ng#1656]
I have a general question about TAGS and a possible feature request.

Several systems in my environment are parsing logs via a pattern database and applying TAGS to events that they process. I would like to use the TAGS in filters on a second set of systems that act as a central collection point for the parsed logs. I have been able to send the TAGS to the second set of systems via an SDATA field, but am unable to use the tags() filter option on the second set of systems. I would like to use tags() in filters on the second set of systems because tags() are significantly more efficient than using string or regex matches in filters.

Is there a dynamic way to set TAGS on the second set of systems (without creating many conditional rewrites or parsing the logs again) or a source option to preserve upstream TAGS?

Either of these would allow one group of systems to do the heavy lifting (parsing, tagging, rewrites, etc) and send the results to another set of systems that would further process the logs according to TAGS.

**Comment by faxm0dem** [comment]
this has popped up several times on the mailing list. My proposal was to have a dedicated `syslog-ng ()` destination that would send all kv pairs and also tags. I totally support this idea

**Comment by bazsi** [comment]
Yeah, absolutely makes sense. I even have a preliminary proof of concept in
this direction. But not enough time on my hands to submit it as a pr.

**Comment by bazsi** [comment]
A small step in this direction is in #1658 which allows you to pass a $TAGS value and parse it back into
message tags.

**Comment by bazsi** [comment]
And here's my preliminary work on creating a "syslog-ng" driver, such that all name-value pairs are transmitted:

**This is the destination driver:**

```
block destination syslog-ng(server('127.0.0.1') transport(tcp) port(514)) {
        network("`server`" transport(`transport`) port(`port`)
                template("<$PRI>1 $ISODATE $LOGHOST @syslog-ng - - ${SDATA:--} $(format-json --scope all-nv-pairs \
                         --exclude 0* --exclude 1* --exclude 2* --exclude 3* --exclude 4* --exclude 5* \
                         --exclude 6* --exclude 7* --exclude 8* --exclude 9* \
                         --exclude SOURCE \
                         --exclude .SDATA.* \
                         ..RSTAMP='${R_UNIXTIME}${R_TZ}' \
                         ..TAGS=${TAGS})\n")
                frac-digits(3)
        );
};
```

**This is the source driver:**

```
block source syslog-ng(transport(tcp) port(514) log-msg-size(65536)) {
        log {
                source {
                        network(transport(`transport`)
                                port(`port`)
                                log-msg-size(`log-msg-size`)
                                flags(syslog-protocol));
                };

                # syslog-ng specific format
                channel {
                        filter { program("@syslog-ng" type(string)); };

                        rewrite {
                                set("$HOST" value("0"));
                                unset(value("PROGRAM"));
                        };
                        # NOTE: this will probably overwrite all builtin
                        # name value pairs, including $MSG
                        parser { json-parser(); };
                        rewrite {
                                # the peer sends its own recvd timestamp
                                # (that we at this point do not process, as
                                # our date-parser() can't parse microsecond
                                # timestamps and is pretty slow.
                                unset(value("..RSTAMP"));

                                # the peer also sends its set of tags, that
                                # we don't process, lacking a tags-parser().
                                unset(value("..TAGS"));

                                set("$(list-append $RELAYS $0)" value("RELAYS"));
                        };
                        flags(final);
                };

                # normal syslog
                channel {
                        flags(final);
                };
        };

};
```

This example has some limitations:
  * it does not process the tags, as the config was made before the tags-parser(), but it is as simple as using parser { tags-parser(template("${..TAGS}")); }; to the appropriate place
  * it is a bit slow, as it won't be able to send more than 20-30k msg/sec, could be enough for your use-case though. Reception is faster.

Any feedback is welcome.

**Comment by faxm0dem** [comment]
looks very good indeed, thanks! I'll test this ASAP.
For the time being, I'd simply add `__VARARGS__` to the blocks

**Comment by Andre188** [comment]
The tags-parser would be a great solution. The event rate numbers on the source seem reasonable as well.

**Comment by bazsi** [comment]
This was implemented in 3.13.2, with the addition of the syslog-ng destination driver and the associated default-network-drivers() source, capable of parsing the format produced by syslog-ng automatically.

Here's a blog post that has some more information on how to use it:
https://www.balabit.com/blog/application-adapters-enterprise-wide-message-model-syslog-ng/
